pkg/networkservice/core/adapters/networkservice_test.go

```diff
@@ -0,0 +1,301 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapters_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/checks/checkcontext"
+)
+
+type contextKeyType string
+
+const contextPassingKey contextKeyType = "context"
+const executionOrderKey contextKeyType = "order"
+
+type appendServer string
+type appendClient string
+
+var (
+	reqCtx = context.WithValue(context.Background(), contextPassingKey, "req")
+	clsCtx = context.WithValue(context.Background(), contextPassingKey, "cls")
+)
+
+func (a appendServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	defer appendContext(&ctx, "+", string(a))()
+	return next.Server(ctx).Request(ctx, request)
+}
+
+func (a appendServer) Close(ctx context.Context, connection *networkservice.Connection) (*empty.Empty, error) {
+	defer appendContext(&ctx, "-", string(a))()
+	return next.Server(ctx).Close(ctx, connection)
+}
+
+func (a appendClient) Request(ctx context.Context, in *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	defer appendContext(&ctx, "+", string(a))()
+	return next.Client(ctx).Request(ctx, in, opts...)
+}
+
+func (a appendClient) Close(ctx context.Context, in *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	defer appendContext(&ctx, "-", string(a))()
+	return next.Client(ctx).Close(ctx, in, opts...)
+}
+
+func appendContext(ctxPtr *context.Context, prefix, str string) func() {
```

Comment:
I think `strings.Builder` could reduce strings copying for these tests.

Reply:
These tests run in `0.038s` on CI, so I think it doesn't worth optimizing.

Reply:
As I can see these test run in 1s on windows 
```
github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters	1.044s
```
It is not super critical but It could be faster.

Reply:
On windows, every test run in at least one second, even such simple as `clienturl`. It looks like every test on windows delayed exactly by 1 second (or time measured incorrectly).